FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+            pxPendingAcks[ i ].packetId = MQTT_PACKET_ID_INVALID;
+            pxPendingAcks[ i ].pCommandContext = NULL;
+            pxPendingAcks[ i ].callback = NULL;
+            break;
+        }
+    }
+
+    return xFoundAck;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvAddSubscription( const char * pTopicFilter,
+                                uint16_t topicFilterLength,
+                                QueueHandle_t pxQueue )
+{
+    int32_t i = 0;
+
+    for( i = 0; i < SUBSCRIPTIONS_MAX_COUNT; i++ )
+    {
+        if( pxSubscriptions[ i ].topicFilterLength == 0 )
+        {
+            pxSubscriptions[ i ].topicFilterLength = topicFilterLength;
+            pxSubscriptions[ i ].pResponseQueue = pxQueue;
+            memcpy( pxSubscriptions[ i ].pTopicFilter, pTopicFilter, topicFilterLength );
+            break;
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvRemoveSubscription( const char * pTopicFilter,
+                                   size_t topicFilterLength,
+                                   QueueHandle_t pxQueue )
+{
+    /* TODO: Unused for now, but can be used to remove a single subscription
+     * when multiple apps are subscribed to the same topic. */
+    ( void ) pxQueue;
+    int32_t i = 0;
+
+    for( i = 0; i < SUBSCRIPTIONS_MAX_COUNT; i++ )
+    {
+        if( pxSubscriptions[ i ].topicFilterLength == topicFilterLength )
+        {
+            if( ( strncmp( pxSubscriptions[ i ].pTopicFilter, pTopicFilter, topicFilterLength ) == 0 ) && true )
+            {
+                pxSubscriptions[ i ].topicFilterLength = 0;
+                pxSubscriptions[ i ].pResponseQueue = NULL;
+                memset( ( void * ) pxSubscriptions[ i ].pTopicFilter, 0x00, sizeof( pxSubscriptions[ i ].pTopicFilter ) );
+                break;
+            }
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvCreateCommand( CommandType_t xCommandType,
+                              CommandContext_t * pxContext,
+                              CommandCallback_t xCallback,
+                              Command_t * pxCommand )
+{
+    bool xIsValid = true;
+
+    memset( ( void * ) pxCommand, 0x00, sizeof( Command_t ) );
+    pxCommand->commandType = xCommandType;
+    pxCommand->pContext = pxContext;
+    pxCommand->callback = xCallback;
+
+    /* Determine if required parameters are present in context. */
+    switch( xCommandType )
+    {
+        case PUBLISH:
+            xIsValid = ( pxContext != NULL ) ? ( pxContext->pPublishInfo != NULL ) : false;
+            break;
+
+        case SUBSCRIBE:
+        case UNSUBSCRIBE:
+            xIsValid = ( pxContext != NULL ) ? ( pxContext->pSubscribeInfo != NULL ) : false;
+            break;
+
+        default:
+            /* Other operations don't need a context. */
+            break;
+    }
```

Comment:
I think it is better to do the validation before updating pxCommand. Is there any reason why you have done this way?

Reply:
No, there isn't. This function originally returned `void` before I decided it would be better to validate it here as well instead of having separate checks. I just happened to add the checks after the assignment but I'll change it so the validation happens first.

Reply:
Thanks